flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java

```diff
@@ -0,0 +1,444 @@
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.junit.Assert.assertThat;
+
+/** Tests for declarative scheduler's {@link Executing} state. */
+public class ExecutingTest extends TestLogger {
+    @Test
+    public void testTransitionToFailing() throws Exception {
+        final String failureMsg = "test exception";
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            ctx.setExpectFailing(
+                    (failingArguments -> {
+                        assertThat(failingArguments.getExecutionGraph(), notNullValue());
+                        assertThat(failingArguments.getFailureCause().getMessage(), is(failureMsg));
+                    }));
+            ctx.setHowToHandleFailure(Executing.FailureResult::canNotRestart);
+            Executing exec = getExecutingState(ctx);
+            exec.onEnter();
+            exec.handleGlobalFailure(new RuntimeException(failureMsg));
+        }
+    }
+
+    @Test
+    public void testTransitionToRestarting() throws Exception {
+        final Duration duration = Duration.ZERO;
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            ctx.setExpectRestarting(
+                    (restartingArguments ->
+                            assertThat(restartingArguments.getBackoffTime(), is(duration))));
+            ctx.setHowToHandleFailure((t) -> Executing.FailureResult.canRestart(duration));
+            Executing exec = getExecutingState(ctx);
+            exec.onEnter();
+            exec.handleGlobalFailure(new RuntimeException("Recoverable error"));
+        }
+    }
+
+    @Test
+    public void testTransitionToCancelling() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            ctx.setExpectCancelling(assertNonNull());
+            Executing exec = getExecutingState(ctx);
+            exec.onEnter();
+            exec.cancel();
+        }
+    }
+
+    @Test
+    public void testTransitionToFinishedOnTerminalState() throws Exception {
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            ctx.setExpectFinished(
+                    archivedExecutionGraph ->
+                            assertThat(archivedExecutionGraph.getState(), is(JobStatus.FAILED)));
+            ctx.setGetMainThreadExecutor(() -> executor);
+            Executing exec = getExecutingState(ctx);
+            ctx.setExpectedStateChecker((state) -> state == exec);
+            exec.onEnter();
+            // transition EG into terminal state, which will notify the Executing state about the
+            // failure (async via the supplied executor)
+            exec.getExecutionGraph().failJob(new RuntimeException("test failure"));
+            executor.shutdown();
+            executor.awaitTermination(10, TimeUnit.SECONDS);
```

Comment:
this should probably be moved into a finally block

Reply:
The problem is that the AutoClosable.close is called before executing the finally block, hence the test fails because the state transition hasn't happened yet.

Reply:
Does that mean we are relying on the executor shutdown to run outstanding tasks? If so, then we should make this more explicit (for example with a `ManuallyTriggeredScheduledExecutorService`) because I did not expect that from a quick glance.

Reply:
Allright, I'll update accordingly.